### PR TITLE
+ group_walk()

### PR DESCRIPTION
**Pull request by romainfrancois**
This needs 
- more work on the `?group_map`
- `group_walk()` example an test

I currently don't have a good example in mind of using `.y` in `group_map()`

### CODE REVIEW

File: R/group_map.R [hunk 1]
```diff
@@ -3,23 +3,37 @@
 #'
 #' \badgeexperimental
 #'
+#' `group_map()` and `group_walk()` are purrr-style function that can
+#' be used to iterate on grouped tibbles.
```
Comment (jennybc) [inline R/group_map.R]: We could place in context (here or somewhere):

  * Use `group_map()` when `summarize()` is too limited, in terms of what you need to do for each group.
  * `group_map()` and `group_walk()` are an evolution of `dplyr::do()`.

File: R/group_map.R [hunk 2]
```diff
@@ -3,23 +3,37 @@
 #'
 #' \badgeexperimental
 #'
+#' `group_map()` and `group_walk()` are purrr-style function that can
```
Comment (jennybc) [inline R/group_map.R]:
```suggestion
#' `group_map()` and `group_walk()` are purrr-style functions that can
```

File: R/group_map.R [hunk 3]
```diff
@@ -3,23 +3,37 @@
...
+#' be used to iterate on grouped tibbles.
+#'
+#' Each conceptual group of the data frame is exposed to the function with two pieces of information:
```
Comment (jennybc) [inline R/group_map.R]:
```suggestion
#' Each conceptual group of the data frame is exposed to the function `.f` with two pieces of information:
```

File: R/group_map.R [hunk 4]
```diff
@@ -3,23 +3,37 @@
...
+#'
+#'   - The subset of the data for the group, exposed as `.x`.
+#'   - The key, a tibble with exactly one row and columns for each grouping variable, exposed as `.y`
```
Comment (jennybc) [inline R/group_map.R]:
```suggestion
#'   - The key, a tibble with exactly one row and columns for each grouping variable, exposed as `.y`.
```

File: R/group_map.R [hunk 5]
```diff
@@ -3,23 +3,37 @@
 #'
 #' \badgeexperimental
```
Comment (jennybc) [inline R/group_map.R]: This has somehow become the entire Description, with everything else showing un in Details.

File: R/group_map.R [hunk 6]
```diff
@@ -3,23 +3,37 @@
...
+#'   - `.y` to refer to the key, a one row tibble with one column per grouping variable
+#'   that identifies the group
 #'
 #' @param ... Additional arguments passed on to `.f`
 #' @param keep Should `.x` contain the grouping variables
 #'
-#' @return The function specified in `.f` is called on each group, and the data frames
-#'         are combined with [bind_rows()]
+#' @return
+#'  - `group_map()` combines the data frames returned by `.f`
```
Comment (jennybc) [inline R/group_map.R]:
```suggestion
#'  - `group_map()` row binds the data frames returned by `.f`
```

File: R/group_map.R [hunk 7]
```diff
@@ -3,23 +3,37 @@
...
+#'  - `group_walk()` calls `.f` for side effects, and eventually invisibly returns `.tbl` unchanged
```
Comment (jennybc) [inline R/group_map.R]:
```suggestion
#'  - `group_walk()` calls `.f` for side effects and returns the input `.tbl`, invisibly
```

File: R/group_map.R [hunk 8]
```diff
@@ -3,23 +3,37 @@
...
 #'
 #' @seealso [group_split()] and [group_keys()]
```
Comment (jennybc) [inline R/group_map.R]: Shouldn't this be handled by ` @family grouping functions`?

**Comment by jennybc** [comment]
More example ideas, including for `group_walk()` and using `.y`:

``` r
## I'm in the group_walk branch
devtools::load_all("~/rrr/dplyr")
#> Loading dplyr

iris %>%
  group_by(Species) %>%
  group_map(~ .x$Petal.Length %>% 
              quantile(probs = c(0.25, 0.5, 0.75)) %>% 
              tibble::enframe(name = "prob", value = "quantile"))
#> # A tibble: 9 x 3
#> # Groups:   Species [3]
#>   Species    prob  quantile
#> * <fct>      <chr>    <dbl>
#> 1 setosa     25%       1.4 
#> 2 setosa     50%       1.5 
#> 3 setosa     75%       1.58
#> 4 versicolor 25%       4   
#> 5 versicolor 50%       4.35
#> 6 versicolor 75%       4.6 
#> 7 virginica  25%       5.1 
#> 8 virginica  50%       5.55
#> 9 virginica  75%       5.88

iris %>%
  group_by(Species) %>%
  group_map(~ .x %>%
              purrr::map_dfc(fivenum) %>%
              mutate(nms = c("min", "Q1", "median", "Q3", "max")))
#> # A tibble: 15 x 6
#> # Groups:   Species [3]
#>    Species    Sepal.Length Sepal.Width Petal.Length Petal.Width nms   
#>  * <fct>             <dbl>       <dbl>        <dbl>       <dbl> <chr> 
#>  1 setosa              4.3         2.3         1            0.1 min   
#>  2 setosa              4.8         3.2         1.4          0.2 Q1    
#>  3 setosa              5           3.4         1.5          0.2 median
#>  4 setosa              5.2         3.7         1.6          0.3 Q3    
#>  5 setosa              5.8         4.4         1.9          0.6 max   
#>  6 versicolor          4.9         2           3            1   min   
#>  7 versicolor          5.6         2.5         4            1.2 Q1    
#>  8 versicolor          5.9         2.8         4.35         1.3 median
#>  9 versicolor          6.3         3           4.6          1.5 Q3    
#> 10 versicolor          7           3.4         5.1          1.8 max   
#> 11 virginica           4.9         2.2         4.5          1.4 min   
#> 12 virginica           6.2         2.8         5.1          1.8 Q1    
#> 13 virginica           6.5         3           5.55         2   median
#> 14 virginica           6.9         3.2         5.9          2.3 Q3    
#> 15 virginica           7.9         3.8         6.9          2.5 max

# there's an unqualified call to `walk2()` somewhere
suppressPackageStartupMessages(library(purrr))

## this still needs tempdir + cleanup treatment, for an official example
iris %>%
  group_by(Species) %>%
  group_walk(~ readr::write_csv(.x, path = paste0(.y$Species, ".csv")))

list.files(pattern = "csv$")
#> [1] "setosa.csv"     "versicolor.csv" "virginica.csv"
```

<sup>Created on 2018-12-06 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1.9000)</sup>

File: R/group_map.R [hunk 9]
```diff
@@ -66,3 +80,21 @@ group_map.grouped_df <- function(.tbl, .f, ..., keep = FALSE) {
...
+
+#' @export
+group_walk.grouped_df <- function(.tbl, .f, ...) {
+  .f <- rlang::as_function(.f)
+
+  # call the function on each group
+  chunks <- group_split(.tbl, keep = isTRUE(keep))
+  keys  <- group_keys(.tbl)
+  group_keys <- map(seq_len(nrow(keys)), function(i) keys[i, , drop = FALSE])
+  walk2(chunks, group_keys, .f)
```
Comment (jennybc) [inline R/group_map.R]: Needs to be `purrr::walk2()` or, probably, added to the purrr compatibility file.

File: R/group_map.R (hunk 5 above)
Comment (romainfrancois) [reply R/group_map.R]: yeah I wonder about that too. @lionel- might have some guidance

File: R/group_map.R (hunk 8 above)
Comment (romainfrancois) [reply R/group_map.R]: That makes sense, I guess the original intent was to have more than this sentence, but I'm starting to think `group_*` could get a small vignette to describe their relationship. @hadley ?

File: R/group_map.R (hunk 5 above)
Comment (lionel-) [reply R/group_map.R]: You have to add `@description` and `@details` sections manually.

**Pull request merged**